src/devices/machine/tsconfdma.cpp

```diff
@@ -0,0 +1,183 @@
+// license:BSD-3-Clause
+// copyright-holders:Andrei I. Holub
+/**********************************************************************
+
+    TS-Conf (ZX-Evolution) DMA Controller
+
+**********************************************************************/
+
+#include "emu.h"
+#include "tsconfdma.h"
+
+#define LOG_TSDMA (1U << 1)
+
+#define VERBOSE (1U << 1)
+#define LOG_OUTPUT_FUNC osd_printf_info
+
+#include "logmacro.h"
+#define LOGDMA(...) LOGMASKED(LOG_TSDMA, __VA_ARGS__)
+
+// device type definition
+DEFINE_DEVICE_TYPE(TSCONF_DMA, tsconfdma_device, "tsconfdma", "TS-Conf DMA Controller")
+
+tsconfdma_device::tsconfdma_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+    : device_t(mconfig, TSCONF_DMA, tag, owner, clock)
+    , m_in_mreq_cb(*this)
+    , m_out_mreq_cb(*this)
+    , m_in_mspi_cb(*this)
+    , m_out_cram_cb(*this)
+{
+}
+
+void tsconfdma_device::device_start()
+{
+    m_in_mreq_cb.resolve_safe(0);
+    m_out_mreq_cb.resolve_safe();
+    m_in_mspi_cb.resolve_safe(0);
+    m_out_cram_cb.resolve_safe();
+
+    save_item(NAME(m_ready));
+    save_item(NAME(m_address_s));
+    save_item(NAME(m_address_d));
+    save_item(NAME(m_block_len));
+    save_item(NAME(m_block_num));
+    save_item(NAME(m_align_s));
+    save_item(NAME(m_align_d));
+    save_item(NAME(m_align));
+}
+
+void tsconfdma_device::device_reset()
+{
+    m_block_num = 0;
+
+    m_ready = true;
+}
+
+bool tsconfdma_device::is_ready()
+{
+    return m_ready;
+}
+
+void tsconfdma_device::set_saddr_l(u8 addr_l)
+{
+    m_address_s = (m_address_s & 0xffffff00) | (addr_l & 0xfe);
+}
+
+void tsconfdma_device::set_saddr_h(u8 addr_h)
+{
+    m_address_s = (m_address_s & 0xffffc0ff) | ((addr_h & 0x3f) << 8);
+}
+
+void tsconfdma_device::set_saddr_x(u8 addr_x)
+{
+    m_address_s = (m_address_s & 0x0003fff) | (addr_x << 14);
+}
+
+void tsconfdma_device::set_daddr_l(u8 addr_l)
+{
+    m_address_d = (m_address_d & 0xffffff00) | (addr_l & 0xfe);
+}
+
+void tsconfdma_device::set_daddr_h(u8 addr_h)
+{
+    m_address_d = (m_address_d & 0xffffc0ff) | ((addr_h & 0x3f) << 8);
+}
+
+void tsconfdma_device::set_daddr_x(u8 addr_x)
+{
+    m_address_d = (m_address_d & 0x0003fff) | (addr_x << 14);
+}
+
+void tsconfdma_device::set_block_len(u8 len)
+{
+    // 1..256
+    m_block_len = len;
+}
+
+void tsconfdma_device::set_block_num_l(u8 num_l)
+{
+    m_block_num = (m_block_num & 0xff00) | num_l;
+}
+
+void tsconfdma_device::set_block_num_h(u8 num_h)
+{
+    // 1..1025 * 2
+    m_block_num = (m_block_num & 0x00ff) | ((num_h & 0x03) << 8);
+}
+
+void tsconfdma_device::start_tx(u8 dev, bool s_align, bool d_align, bool align_opt)
+{
+    m_ready = false;
+    m_align = align_opt ? 512 : 256;
+
+    // TODO Transfers 2 byte/cyrcle at 7MHz
+    if (dev == 0b0001)
+    // Mem -> Mem
```

Comment:
Please put comments like this on the end of the `if` line or inside the block – it makes things more readable.

But really, this whole `if`/`else` tree would be more readable as a `switch` – it’s only testing for specific values.